### ISSUE TITLE
Output rendering isn't smooth, split-second screen cheese

### ISSUE THREAD

**Issue by jrieken** [microsoft/vscode#116786]
* open GH issue notebook
* repeatedly run the same cell
* when output comes it there is flicker (see attached recording and capture)

<img width="703" alt="Screenshot 2021-02-16 at 16 47 24" src="https://user-images.githubusercontent.com/1794099/108086601-d9f38880-7076-11eb-8c0d-cb35329b1e27.png">
(from the recording)

https://user-images.githubusercontent.com/1794099/108086624-deb83c80-7076-11eb-9855-c0d27cf61817.mov

**Comment by roblourens** [comment]
From memory, I think we

- Render the output
- Get its size
- Tell list to do layout
- scheduleAtNextAnimationFrame
- Apply the layout

Either we have to get rid of waiting a frame which I don't think we can, or render the output to be cut off outside of the space that it previously had.

**Comment by jrieken** [comment]
Unsure if that's the same issue or not but the same visual stutter is true for rich diff editors

<img width="540" alt="Screenshot 2021-02-17 at 13 12 59" src="https://user-images.githubusercontent.com/1794099/108202740-e385fa80-7121-11eb-8146-70520c23ace8.png">

https://user-images.githubusercontent.com/1794099/108202755-e5e85480-7121-11eb-89a5-828a7f66afb1.mov

**Comment by rebornix** [comment]
> Tell list to do layout
> scheduleAtNextAnimationFrame
> Apply the layout
> 
> Either we have to get rid of waiting a frame which I don't think we can

we can't if the list view is in its rendering cycle but we can revisit how bad it is.

**Comment by rebornix** [comment]
Pushed a fix/workaround https://github.com/microsoft/vscode/commit/482013c77e247d1704e7071522dd4f4edea8c128 for this.

The problem is not the `scheduleAtNextAnimationFrame`. When we rerun the cell, we will get a new output so what happens then

1. remove the old output, which will set the output height back to 0
2. send the new output rendering request to the webview
3. output renderer handles the request, render the output
4. browser does its rendering
5. resize listener detects height change, broadcast it to the main frame
6. we adjust the top of list view items

The new output rendered in the webview overlaps with next cell from step 3 to 5, it can be tens of milliseconds then our eyes will not notice, or it might be 500 milliseconds and you can tell easily.

To make our eyes more comfortable, we can prevent the output container from shrinking before step 5. The catch is there is no event from the browser telling us when it finishes the rendering of an HTML fragment (so we don't know precisely when step 5 ends). Right now in the fix, I set it to one second, then if the rendering finishes with one second, you will not see overlap. One second is also not too long otherwise you will see the output container occupies whitespace for quite a while.

The fix will only trigger when replacing outputs, clearing outputs will adjust the height instantly as removing DOM nodes are fast.

**Comment by roblourens** [comment]
I don't totally understand the fix. Should it fix the issue in general or make it less noticeable? It seems like this can still happen


https://user-images.githubusercontent.com/323878/108894815-a7daad00-75c7-11eb-861b-5edacb995707.mp4

Screenshot of the overflow frame

![image](https://user-images.githubusercontent.com/323878/108894877-bcb74080-75c7-11eb-952a-4b20b716e02a.png)

**Comment by rebornix** [comment]
@roblourens it can still happen but only making it less noticeable. We can play around with the delay (right now 1 second) if ppl are still seeing it often.

**Comment by roblourens** [comment]
> if the rendering finishes with one second, you will not see overlap

I don't understand, in my case it finished in less than one second, still overlapped, and is noticeable.

**Comment by jrieken** [comment]
I tend to agree. The original issue is much better but I'd that rendering isn't smooth yet, there is occasional background flashing and awkward resizing... Adding `verified` to this one but creating a new issue for a concrete case that I have recorded
